Core/Entities: add DB-support for conversation actorGuids

funjoker opened the pull request:
<!--- (**********************************)
      (** Fill in the following fields **)
      (**********************************) --->

**Changes proposed:**

-  Added the option to add permanent spawns
-  Use only for "creature_text" like conversations.

**Target branch(es):** 3.3.5/master

- [ ] 3.3.5
- [x] master

**Issues addressed:** 


**Tests performed:** (Does it build, tested in-game, etc.)
Tested Malfurion Ysera conversation in Val'sharah, with sniffed conversation data.

![wowscrnshot_103117_041118](https://user-images.githubusercontent.com/846960/32205861-0c34e31a-bdf2-11e7-97f3-94a1c2fe8b52.jpg)
![wowscrnshot_103117_041125](https://user-images.githubusercontent.com/846960/32205862-0c523578-bdf2-11e7-89f2-0c96bf13d720.jpg)
![wowscrnshot_103117_041132](https://user-images.githubusercontent.com/846960/32205863-0c7f37da-bdf2-11e7-8a66-dd0f8e398a27.jpg)
![wowscrnshot_103117_041143](https://user-images.githubusercontent.com/846960/32205864-0c9d5256-bdf2-11e7-906b-99aef68a7bf7.jpg)

**Known issues and TODO list:** (add/remove lines as needed)

- [ ] Whatever TC wants
- [ ] 


<!--- Notes
- Enable the setting "[√] Allow edits from maintainers." when creating your pull request.
- If this PR only contains SQL files, open a new issue instead and post or link the SQL in the issue.
- When adding new SQL files, name them 9999_99_99_99_db_name.sql to reduce the chance of possible merge conflicts.
--->

funjoker committed:
Add DB-support for conversation actorGuids

funjoker committed:
Tabfix

joschiwald commented on src/server/game/Entities/Conversation/Conversation.cpp:
```diff
@@ -130,7 +131,22 @@ bool Conversation::Create(ObjectGuid::LowType lowGuid, uint32 conversationEntry,
             SetDynamicStructuredValue(CONVERSATION_DYNAMIC_FIELD_ACTORS, actorsIndex++, &actorField);
         }
-        else
+        else if (!conversationTemplate->Actors.empty())
```
this won't work if you have a mix of both actor types, you need to start from 0 when adding actors with guid

funjoker replied:
you are right. Gonna fix that

joschiwald commented on src/server/game/Globals/ConversationDataStore.cpp:
```diff
@@ -103,18 +105,37 @@ void ConversationDataStore::LoadConversationTemplates()
                     actors.resize(idx + 1);
                 actors[idx] = conversationActorTemplate;
-                ++count;
+                ++count;                
```
whitespace

joschiwald commented on src/server/game/Globals/ConversationDataStore.cpp:
```diff
@@ -103,18 +105,37 @@ void ConversationDataStore::LoadConversationTemplates()
             if (ConversationActorTemplate const* conversationActorTemplate = Trinity::Containers::MapGetValuePtr(_conversationActorTemplateStore, actorId))
             {
+                if (actorGuid != 0)
+                    TC_LOG_ERROR("sql.sql", "Table `conversation_actors` references both actor (ID: %u) and actorGuid (GUID: " UI64FMTD ") for Conversation %u, GUID is ignored", actorId, actorGuid, conversationId);
```
move this message above the if condition, and check what your message says

funjoker replied:
It shall just print this error.
Due to other if clause the guid is ignored, See some lines below.

joschiwald commented on src/server/game/Globals/ConversationDataStore.cpp:
```diff
@@ -103,18 +105,37 @@ void ConversationDataStore::LoadConversationTemplates()
+                ++count;                
             }
             else
                 TC_LOG_ERROR("sql.sql", "Table `conversation_actors` references an invalid actor (ID: %u) for Conversation %u, skipped", actorId, conversationId);
```
you will print this error for all actors with guid

funjoker replied:
will adjust all Logs when at home thank you

Vincent-Michael commented on src/server/game/Entities/Conversation/Conversation.cpp:
```diff
@@ -19,6 +19,7 @@
 #include "Map.h"
 #include "Unit.h"
 #include "UpdateData.h"
+#include "Creature.h"
```
really need?

funjoker replied:
bounds.first->second->GetGUID()

Shauren replied:
sort includes alphabetically, keeping `Conversation.h` first

joschiwald commented on src/server/game/Globals/ConversationDataStore.cpp:
```diff
@@ -103,18 +105,37 @@ void ConversationDataStore::LoadConversationTemplates()
             uint32 conversationId = fields[0].GetUInt32();
             uint32 actorId = fields[1].GetUInt32();
-            uint16 idx = fields[2].GetUInt16();
+            uint64 actorGuid = fields[2].GetUInt64();
```
ObjectGuid::LowType

funjoker replied:
Ok

joschiwald commented on src/server/game/Globals/ConversationDataStore.cpp:
```diff
@@ -103,18 +105,37 @@ void ConversationDataStore::LoadConversationTemplates()
+                    ++count;
+                }
+                else
+                    TC_LOG_ERROR("sql.sql","Table `conversation_actors` references an invalid creature guid (GUID: " UI64FMTD ") for Conversation %u, skipped", actorGuid, conversationId);
```
missing space after comma

Shauren commented on src/server/game/Entities/Conversation/Conversation.cpp:
```diff
@@ -130,7 +131,22 @@ bool Conversation::Create(ObjectGuid::LowType lowGuid, uint32 conversationEntry,
+        {
+            auto bounds = map->GetCreatureBySpawnIdStore().equal_range(actorGuids);
+
+            if (bounds.first == bounds.second)
```
This isnt (yet) part of master branch but having dynamic spawning in mind, there could be multiple spawns for the same db guid and you probably should use the alive/last one

Shauren commented on src/server/game/Globals/ConversationDataStore.cpp:
```diff
@@ -21,6 +21,7 @@
 #include "DB2Stores.h"
 #include "Log.h"
 #include "Timer.h"
+#include "ObjectMgr.h"
```
sort

funjoker committed:
Requested changes

funjoker commented on src/server/game/Entities/Conversation/Conversation.cpp:
```diff
@@ -134,6 +136,28 @@ bool Conversation::Create(ObjectGuid::LowType lowGuid, uint32 conversationEntry,
+        if (actorGuids)
+        {
+            auto bounds = map->GetCreatureBySpawnIdStore().equal_range(actorGuids);
+
```
What is the best way to get the creature, checking it's alive, compatible with the dynamic_spawn system ?

funjoker committed:
Better loop

funjoker commented:
#20758 
contains checks if actors are fully set.

Both together will make it fit.

funjoker committed:
Typo

funjoker committed:
Typo

funjoker committed:
fix

funjoker committed:
Move check from ConversationScript PR to this

joschiwald commented on sql/updates/world/master/9999_99_99_99_world.sql:
```diff
@@ -0,0 +1,5 @@
+	CHANGE COLUMN `ConversationActorId` `ConversationActorId` INT(10) UNSIGNED NOT NULL DEFAULT '0' AFTER `ConversationId`,
+	ADD COLUMN `ConversationActorGuid` BIGINT(20) UNSIGNED NOT NULL DEFAULT '0' AFTER `ConversationActorId`,
+	DROP PRIMARY KEY,
+	ADD PRIMARY KEY (`ConversationId`, `ConversationActorId`, `ConversationActorGuid`);
```
what about pk with `ConversationId` and `Idx`?

joschiwald committed:
validate actor fields correctly

Commit:
Typo correction

Overlooked typo (should have been changed when the other `actorsIndex` entries were renamed).

joschiwald committed:
added missing log message
